### PR TITLE
all: Add support for version command

### PR DESCRIPTION
**Pull request by Jordan9500**
When doing version it will display the commit and version 
Update gitignore with the new files generated by the make 
Update the display for when the args are less than 2 
Use a util file to store the Version and Commit Hash

Fixes: #477
Signed-off-by: Jordan Jackson <jordan.jackson@ibm.com>

### CODE REVIEW

**Comment by Jordan9500** [comment]
With the changes being made in the refactor PR, I will wait before making further changes for that to be merged

Depends on: https://github.com/confidential-containers/cloud-api-adaptor/pull/329

**Review by yoheiueda — COMMENTED** [review]
@Jordan9500 Could you update this PR if possible, since #329 has been merged.

Only providers IBM Cloud and AWS have been switched to the new code layout. It is OK If you want to wait until all of the providers need to be refactored.

**Comment by Jordan9500** [comment]
I have updated the code with the latest changes. My plan with the version check was to implement it like: 
```
/usr/local/bin/cloud-api-adaptor version 
cloud-api-adaptor, version: unknown, commit: <current-commit-hash>
go: <go version>
```
I originally thought that adding the code to the top of the function using os.Args that was used before would allow the use across all providers. (Not just the ones moved to the new refactor)

In regards to the current failures, i'm looking into a different way that does not use the make file to generate a version file that gets used. Or atleast uses a file that has the variables defined as `unknown`.

File: cmd/parse.go [hunk 1]
```diff
@@ -9,39 +9,21 @@ import (
...
-			Exit(1)
-		}
-
+	if len(args) == 0 {
+		flags.PrintDefaults()
+		Exit(1)
 	}
 
-	if err := flags.Parse(args[1:]); err != nil {
+	if err := flags.Parse(args); err != nil {
```
Comment (yoheiueda) [inline cmd/parse.go:22]: Please keep to use `args[1:]` to suppress the unit test failures. `flags.Parse` assumes `args[0]` is a program name.

**Review by yoheiueda — CHANGES_REQUESTED** [review]
Could you use the `-ldflags`technique to embed a version information in a binary, instead of modifying `version.go`? Modifying a git-managed file may cause confusion.

https://www.digitalocean.com/community/tutorials/using-ldflags-to-set-version-information-for-go-applications

In the future, we need to switch to a modern third-party command line parser library. I originally implemented this simple parser, but I don't think it is suitable for the current complicated scenario.

**Comment by yoheiueda** [comment]
Thank you for updating the patch.

The second commit messages needs to be formatted as your first commit message. Could you update your commit message? Note that at least one of commits needs to have a `Fixes` line.

**Comment by Jordan9500** [comment]
I have updated the PR to match the changes you have requested, and got the ldflags working. 
This is now the output, without the generated file steps from the previous commit.
```
/usr/local/bin/cloud-api-adaptor version
cloud-api-adaptor, version: unknown, commit: 3afd0f5bd1bc338dc44213ef300ed3bc05a00b32
go: go1.18.5
```

**Comment by yoheiueda** [comment]
@Jordan9500 could you rebase to resolve the conflicts in main.go? Cloud providers for azure and libvirt have been merged.

**Comment by Jordan9500** [comment]
I have rebased with the latest changes

**Review by yoheiueda — APPROVED** [review]
LGTM. Thank you!

**Review by bpradipt — APPROVED** [review]
/lgtm
Thanks @Jordan9500

File: cmd/cloud-api-adaptor/main.go [hunk 2]
```diff
@@ -35,10 +35,15 @@ type networkConfig struct {
 func (cfg *daemonConfig) Setup() (cmd.Starter, error) {
 
 	if len(os.Args) < 2 {
-		fmt.Printf("%s aws|azure|ibmcloud|libvirt|vsphere <options>\n", os.Args[0])
+		fmt.Printf("%s \n<aws|azure|ibmcloud|libvirt|version> [options]\n", os.Args[0])
```
Comment (stevenhorsman) [inline cmd/cloud-api-adaptor/main.go:38]: @Jordan9500 - did you mean to remove vsphere here?

Comment (Jordan9500) [reply cmd/cloud-api-adaptor/main.go:38]: Good spot, I did not mean to remove that from the print. 
I will make another PR to re add that command to the print. 
Thank you :)

**Pull request merged**